src/QsCompiler/CompilerOptimizations/FunctionEvaluation.fs

```diff
@@ -6,153 +6,147 @@ open Microsoft.Quantum.QsCompiler.DataTypes
+type private FunctionInterrupt =
 | Returned of Expr // The function is returning a value
 | Failed of Expr // The function is throwing an exception
 | CouldNotEvaluate of string // I couldn't evaluate the function
-
-/// The current state of an expression evaluation
-type private ExpressionResult<'A> =
-| ExprValue of 'A
-| ExprError of string
+/// The current state of a function evaluation
+type private FunctionState = Result<TransformationState, FunctionInterrupt>
 
 
 /// Evaluates functions by stepping through their code
-type [<AbstractClass>] internal FunctionEvaluator(cd: CallableDict) =
+type [<AbstractClass>] internal FunctionEvaluator() =
 
     /// Transforms a BoolLiteral into the corresponding bool
     let castToBool x =
         match x with
-        | BoolLiteral b -> ExprValue b
-        | _ -> "Not a BoolLiteral: " + (printExpr x) |> ExprError
-
-    /// Returns whether a FunctionState is an interrupt
-    let isInterrupt = function Normal -> false | _ -> true
+        | BoolLiteral b -> Ok b
+        | _ -> "Not a BoolLiteral: " + (printExpr x) |> CouldNotEvaluate |> Error
 
     /// The callback for the subroutine that evaluates and simplifies an expression
-    abstract member evaluateExpression: VariablesDict -> TypedExpression -> Expr
+    abstract member evaluateExpression: TransformationState -> TypedExpression -> Expr
 
     /// Evaluates a single Q# statement
-    member private this.evaluateStatement (vars: VariablesDict) (statement: QsStatement): FunctionState =
+    member private this.evaluateStatement (state: TransformationState) (statement: QsStatement): FunctionState =
         match statement.Statement with
         | QsExpressionStatement expr ->
-            this.evaluateExpression vars expr |> ignore; Normal
+            this.evaluateExpression state expr |> ignore; state |> Ok
         | QsReturnStatement expr ->
-            this.evaluateExpression vars expr |> Returned
+            this.evaluateExpression state expr |> Returned |> Error
         | QsFailStatement expr ->
-            this.evaluateExpression vars expr |> Failed
+            this.evaluateExpression state expr |> Failed |> Error
         | QsVariableDeclaration s ->
-            fillVars vars (StringTuple.fromSymbolTuple s.Lhs, this.evaluateExpression vars s.Rhs); Normal
+            defineVarTuple state (s.Lhs, this.evaluateExpression state s.Rhs) |> Ok
         | QsValueUpdate s ->
             match s.Lhs.Expression with
-                | Identifier (LocalVariable name, tArgs) -> vars.setVar(name.Value, this.evaluateExpression vars s.Rhs); Normal
-                // TODO - allow any symbol tuple on the LHS
-                | _ -> "Unknown LHS of value update statement: " + (printExpr s.Lhs.Expression) |> CouldNotEvaluate
+            | Identifier (LocalVariable name, tArgs) -> setVar state (name.Value, this.evaluateExpression state s.Rhs) |> Ok
+            // TODO - allow any symbol tuple on the LHS
```

Comment:
How about implementing this TODO?

Reply:
I will eventually, but I had some questions - why is the LHS an expression, and not a SymbolTuple as the LHS of a variable declaration? Could I handle these in the same way?

Reply:
Keeping the lhs an expression seemed more future proove. Right now, the expression on the lhs necessarily can only contain (mutable) symbols and tuples thereof - and missing expressions.

Reply:
Fixed in future PR